### PR TITLE
Improve Application Insights exporter

### PR DESCRIPTION
**Pull request by SergeyKanzhelev**
Conversion from SpanData to Application Insights was already implemented and tested here: https://github.com/Microsoft/ApplicationInsights-LocalForwarder/ The idea is to get an exporter in this repository to parity and then replace logic in Local Forwarder with this implementation.

First test is passing - opening PR for initial feedback.

CC: @lmolkova

### CODE REVIEW

File: test/OpenCensus.Exporter.ApplicationInsights.Tests/Implementation/TraceExporterHandlerTests.cs [hunk 1]
```diff
@@ -0,0 +1,1949 @@
...
+            Assert.Equal(url.LocalPath, dependency.Data);
+            Assert.Equal("POST /path", dependency.Name);
+            Assert.Equal("200", dependency.ResultCode);
+            Assert.Null(dependency.Target);
+            Assert.Equal("Http", dependency.Type);
+        }
+
+        /*
+        [Fact]
+        public void OpenCensusTelemetryConverterTests_TracksHttpDependencyWithUrlIgnoresHostPortPath()
```
Comment (lmolkova) [inline test/OpenCensus.Exporter.ApplicationInsights.Tests/Implementation/TraceExporterHandlerTests.cs]: why so many tests are commented out?

File: src/OpenCensus.Exporter.ApplicationInsights/Implementation/TraceExporterHandler.cs [hunk 2]
```diff
@@ -38,38 +39,158 @@ public void Export(IList<ISpanData> spanDataList)
...
-                        (s) => { return s; },
-                        (b) => { return b.ToString(); },
-                        (l) => { return l.ToString(); },
-                        (d) => { return d.ToString(); },
-                        (obj) => { return obj.ToString(); });
+                    result.Context.Operation.ParentId = string.Concat("|", traceId, ".", parentId, ".");
+                }
+
+                // TODO: I don't understant why this concatanation is required
+                result.Id = string.Concat("|", traceId, ".", spanId, ".");
```
Comment (lmolkova) [inline src/OpenCensus.Exporter.ApplicationInsights/Implementation/TraceExporterHandler.cs]: this is not required in a bright future, but required now for back-compatibility with AppInsights (assuming AI SDK talk to opencensus). We can talk about it and plan to remove it, but for now this is required.

File: src/OpenCensus.Exporter.ApplicationInsights/Implementation/TraceExporterHandler.cs [hunk 3]
```diff
@@ -82,5 +203,233 @@ public void Export(IList<ISpanData> spanDataList)
...
+                }
+            }
+
+            // 1 tick is 100 ns
+            timestamp = DateTimeOffset.FromUnixTimeSeconds(span.StartTimestamp.Seconds);
+            timestamp = timestamp.Add(TimeSpan.FromTicks(span.StartTimestamp.Nanos / 100));
+
+            name = span.Name;
+
+            props = new Dictionary<string, string>();
```
Comment (lmolkova) [inline src/OpenCensus.Exporter.ApplicationInsights/Implementation/TraceExporterHandler.cs]: could you avoid creating a new dictionary and copying properties? instead just pass telemetry properties here by ref?

File: src/OpenCensus.Exporter.ApplicationInsights/Implementation/TraceExporterHandler.cs [hunk 4]
```diff
@@ -38,38 +39,158 @@ public void Export(IList<ISpanData> spanDataList)
...
+                            (l) => { return l.ToString(); },
+                            (d) => { return d.ToString(); },
+                            (obj) => { return obj.ToString(); });
+
+                        props.Add(attr.Key, value);
+                    }
+                }
 
-                if (span.Kind == SpanKind.Server)
+                this.OverwriteSpanKindFromAttribute(spanKindAttr, ref resultKind);
```
Comment (lmolkova) [inline src/OpenCensus.Exporter.ApplicationInsights/Implementation/TraceExporterHandler.cs]: most likely it's never needed. How about we do it only if there is no span kind set on the span?

File: src/OpenCensus.Exporter.ApplicationInsights/Implementation/TraceExporterHandler.cs [hunk 5]
```diff
@@ -38,38 +39,158 @@ public void Export(IList<ISpanData> spanDataList)
...
+                    else
+                    {
+                        var value = attr.Value.Match<string>(
+                            (s) => { return s; },
+                            (b) => { return b.ToString(); },
+                            (l) => { return l.ToString(); },
+                            (d) => { return d.ToString(); },
+                            (obj) => { return obj.ToString(); });
+
+                        props.Add(attr.Key, value);
```
Comment (lmolkova) [inline src/OpenCensus.Exporter.ApplicationInsights/Implementation/TraceExporterHandler.cs]: always expect duplicates - we don't want it to throw

File: src/OpenCensus.Exporter.ApplicationInsights/Implementation/TraceExporterHandler.cs [hunk 6]
```diff
@@ -38,38 +39,158 @@ public void Export(IList<ISpanData> spanDataList)
...
+                IAttributeValue httpHostAttr = null;
+                IAttributeValue httpUrlAttr = null;
+                IAttributeValue httpUserAgentAttr = null;
+                IAttributeValue httpRouteAttr = null;
+                IAttributeValue httpPortAttr = null;
+
+                foreach (var attr in span.Attributes.AttributeMap)
+                {
+                    var key = attr.Key;
+                    if (key == "span.kind")
```
Comment (lmolkova) [inline src/OpenCensus.Exporter.ApplicationInsights/Implementation/TraceExporterHandler.cs]: nit: use new C# pattern matching for switch with a string like  

```csharp
switch (str)
       {
          case "1":
          case "small":
              cost += 25;
              break;
          case "2":
          case "medium":
              cost += 25;
              goto case "1";
          case "3":
          case "large":
              cost += 50;
              goto case "1";
          default:
              Console.WriteLine("Invalid selection. Please select 1, 2, or 3.");
              break;
      }
```

File: test/OpenCensus.Exporter.ApplicationInsights.Tests/Implementation/TraceExporterHandlerTests.cs (hunk 1 above)
Comment (SergeyKanzhelev) [reply test/OpenCensus.Exporter.ApplicationInsights.Tests/Implementation/TraceExporterHandlerTests.cs]: Fixed most of them. Tests with the node are still commented out - we need some Node API

File: src/OpenCensus.Exporter.ApplicationInsights/Implementation/TraceExporterHandler.cs (hunk 3 above)
Comment (SergeyKanzhelev) [reply src/OpenCensus.Exporter.ApplicationInsights/Implementation/TraceExporterHandler.cs]: It's harder here. I will think how to restructure code later alongside other perf improvements

**Pull request merged**